File: leet-code/004-MedianOfTwoSortedArrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int pos1 = (nums1.size() + nums2.size() + 1) / 2 - 1;
        int pos2 = (nums1.size() + nums2.size()) / 2;
        int i = 0, j = 0, k = 0;
        int m1 = -1, m2 = -1;
        while((i < nums1.size() || j < nums2.size()) && k <= pos2) {
            int m = -1;
            if(i < nums1.size() && j < nums2.size()) {
                if(nums1[i] <= nums2[j]) {
                    m = nums1[i];
                    i++;
                } else {
                    m = nums2[j];
                    j++;
                }
            } else if(i < nums1.size()) {
                m = nums1[i];
                i++;
            } else {
                m = nums2[j];
                j++;
            }
            if(k == pos1) {
                m1 = m;
            }
            if(k == pos2) {
                m2 = m;
            }
            k++;
        }
        return m1 + (m2-m1)/2.0;
    }
};

File: leet-code/004-MedianOfTwoSortedArrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int p = nums1.length;
        int q = nums2.length;
        int m1 = (p + q + 1) / 2 - 1;
        int m2 = (p + q) / 2;
        int i = 0, j = 0, k = 0, m = 0;
        while (k <= m2) {
            int x = 0;
            if (i < p) {
                if (j < q) {
                    if (nums1[i] <= nums2[j]) {
                        x = nums1[i++];
                    } else {
                        x = nums2[j++];
                    }
                } else {
                    x = nums1[i++];
                }
            } else {
                x = nums2[j++];
            }
            if (k == m1) m += x;
            if (k == m2) m += x;
            k++;
        }
        return m / 2.0;
    }
}
